doc: added a new topic: Compiled class hash (#1294)
starknet-io/starknet-docs: components/Starknet/modules/architecture-and-concepts/pages/smart-contracts/compiled_class_hash.adoc

Reviewer: @raizo07 I know this is annoying, but please change the file name to use hyphens instead of underscores. Search engines interpret words connected by underscores as single words, but they view hyphens as separating words.

So the file name should be `compiled-class-hash.adoc`.

[id="compiled_class_hash"]
= Compiled class hash

> The compiled class hash is a cryptographic hash that results from the compilation process of a Cairo class

Reviewer: @raizo07  Can you start the topic by stating why the reader should care about it? Is it something that they will use in some way? If not, then who does use it and for what purpose?

from its intermediate representation (Sierra) to Cairo assembly (Casm). This process is managed by the Sierra→Casm compiler.

> The compiled class hash is a crucial component within the Starknet ecosystem, specifically used to uniquely identify the output generated by the Sierra to Casm (Cairo assembly) compiler. This hash ensures that each compiled class output is distinct and verifiable.

Reviewer: > specifically used to uniquely identify the output generated by the Sierra to Casm (Cairo assembly) compiler

Who/what uses it?

> The compiled class hash is a crucial component within the Starknet ecosystem, specifically used to uniquely identify the output generated by the Sierra to Casm (Cairo assembly) compiler. This hash ensures that each compiled class output is distinct and verifiable.

Reviewer: ```suggestion
The compiled class hash is a crucial component within the Starknet ecosystem, specifically used to uniquely identify the output generated by the Sierra to Casm compiler. This hash ensures that each compiled class output is distinct and verifiable.
```

> Within Starknet, Cairo classes forming part of the state commitment are specified with Sierra, acting as an intermediate representation bridging Cairo and Cairo assembly (Casm). Despite this, provers operate solely with Casm.

Reviewer: ```suggestion
The state commitment uses the Sierra code that results when compiling Cairo classes. Sierra acts as an intermediate representation between Cairo and Casm. Provers, however, operate solely with Casm.
```

> To circumvent the necessity of recompiling from Sierra to Casm each block in which the class is deployed, the state commitment necessitates relevant information concerning the corresponding Cairo assembly. This requisite information is encapsulated within the compiled class hash.

Reviewer: > To circumvent the necessity of recompiling from Sierra to Casm each block in which the class is deployed, the state commitment necessitates relevant information concerning the corresponding Cairo assembly. This requisite information is encapsulated within the compiled class hash.

Please use simpler language. In general, use simple words when possible. E.g. _needs_ instead of _necessitates_, _contained_ instead of _encapsulated_, etc.

> To circumvent the necessity of recompiling from Sierra to Casm each block in which the class is deployed, the state commitment necessitates relevant information concerning the corresponding Cairo assembly. This requisite information is encapsulated within the compiled class hash.

Reviewer: ```suggestion
In order to avoid recompiling, from Sierra to Casm, each block in which the class is deployed, the state commitment gets the information it needs about the corresponding Casm from the the information contained in the compiled class hash.
```

> When declaring a contract, the party administering the contract endorses the compiled class hash, procured using an SDK, as an integral component of the xref:Network_Architecture/transactions.adoc#declare_v2[`DECLARE`] transaction. Following the inclusion of the transaction in a block, the compiled class hash integrates into the state commitment.

Reviewer: ```suggestion
When declaring a contract, the party administering the contract endorses the compiled class hash, procured using an SDK, as an integral component of the xref:network-architecture/transactions.adoc#declare_v2[`DECLARE`] transaction. Following the inclusion of the transaction in a block, the compiled class hash integrates into the state commitment.
```

> == Purpose and Significance

Reviewer: This section restates a lot of what was stated above, but it's stated more clearly and succinctly here.
I suggest taking my comments above, and then see if you can prune out the redundant info, but use this clear and succinct way of explaining how the hash is used towards the beginning of the topic.

* Uniqueness: The compiled class hash ensures the uniqueness of each compiled class. It is essentially a fingerprint for the compiled output, allowing the network to verify the integrity and uniqueness of the class.

> * State Commitment: In Starknet, state commitment includes various components, one of which is the Cairo classes. These classes are initially defined using Sierra. However, for the prover (the entity responsible for generating cryptographic proofs) to function efficiently, it requires Casm, the final compiled form of these classes.

Reviewer: ```suggestion
* State Commitment: In Starknet, state commitment includes various components, including the Cairo classes. These classes are initially defined using Sierra. However, for the prover to function efficiently, it requires Casm.
```

== Usage

When a new contract is declared on Starknet, the compiled class hash plays a pivotal role. Here's how:

* Declaration Process: The party declaring the contract computes the compiled class hash using an SDK provided by Starknet.

* DECLARE Transaction: This hash is then included as part of the xref:Network_Architecture/transactions.adoc#declare_v2[`DECLARE`] transaction. The xref:Network_Architecture/transactions.adoc#declare_v2[`DECLARE`] transaction is a specific type of transaction in Starknet used to register new contracts.

* Inclusion in State Commitment: Once the xref:Network_Architecture/transactions.adoc#declare_v2[`DECLARE`] transaction is included in a block, the compiled class hash becomes part of the state commitment. This inclusion ensures that the network recognizes and stores the unique compiled output of the contract.

Prospectively, as Sierra-to-Casm compilation integrates into the Starknet OS, the value might undergo updates via proof of the Sierra-to-Casm compiler execution. Such verification demonstrates that compiling the same class with an updated compiler version yields a fresh compiled class hash.

> In summary, the compiled class hash is a foundational element in Starknet’s architecture, enabling efficient state commitment and ensuring the integrity and uniqueness of compiled classes.

Reviewer: ```suggestion
The compiled class hash is a basic element in Starknet’s architecture, enabling efficient state commitment and ensuring the integrity and uniqueness of compiled classes.
```
